### PR TITLE
Add include() function

### PR DESCRIPTION
**Pull request by TheQwertiest**
Based on https://github.com/mesonbuild/meson/pull/3557.

Proposal: global `include(path_to_file)` function that works similar to `subdir(dir)`, but has the following restrictions:
 * The path can be either relative or absolute and can point to the arbitrary file (if it satisfies restrictions listed below).
 * Included file must be a valid meson file.
 * Same file can be included only once.
 * Can't include `meson.build` file.
 * Can't include file from outside of source directory.
 * Can't include file from subprojects directory (unless invoked from inside subprojects directory).
 * File must have a specific extension (currently `*.meson`).
 
Reasoning:
The main purpose of `include()` is to avoid overly long `meson.build` by splitting it in multiple files and to avoid needless clutter (currently you have to create a separate folder with `meson.build` inside to achieve the same functionality).

### CODE REVIEW

**Comment by TheQwertiest** [comment]
I would also like to require some specific suffix\prefix (reasoning https://github.com/mesonbuild/meson/pull/3557#issuecomment-467416749, https://github.com/mesonbuild/meson/pull/3557#issuecomment-467417429), but I have no naming sense, so any help would be greatly appreciated =)
Possible suffixes/prefixes:
```
*.meson
*.meson.build
*.meson.include
meson.*
meson_*
```

**Comment by TheQwertiest** [comment]
@xclaesse, since you were for (or rather not against) the previous `include` proposal, could you, please, take a look at this one as well? =)

**Comment by jpakkane** [comment]
One **big** problem with this is that there are many places in the code that assume that things are defined in files called `meson.build`. This will cause all sorts of unpleasant breakage and regressions. See #5048 for just a taste of it.

**Comment by TheQwertiest** [comment]
Hm... So, I take it you are not against the idea of this PR? =)
I can try tracking down `meson.build` usage and see if it's possible to avoid such breakage.

**Comment by dcbaker** [comment]
I'm going to admit that I'm super not excited about this. One of the things that makes understanding CMake build systems awful is that they have this pattern of shoving all kinds of useful information into files in cmake/ that need a flow chart to understand the interactions between them. The other concern I have is that this could be used to (poorly) implement functions and break the not Turing complete guarantee in the worst way possible, like:

```meson
if something:
  include('foo.meson')
```
foo.meson:
```meson
if something:
  include('foo.meson')
```
Even if we specifically guard against recursion, you still effectively have macros.

**Comment by rilian-la-te** [comment]
@dcbaker there is a goal to avoid all kinds of macro in meson? (even if macro did not make meson Turing-compatible)?

**Comment by TheQwertiest** [comment]
>  they have this pattern of shoving all kinds of useful information into files in cmake/ that need a flow chart to understand the interactions between them

Alternatives currently available in `meson` are:
1. One big `meson.build` (I've encountered several 5k LOC ones).
2. Splitting `meson.build` by creating multiple nested `fake` subdirs (each one contains only a single `meson.build`).

And I'm really not sure if these alternatives are any better or easier to understand: IMO, alternative 1. is unmaintainable mess; while alternative 2. is mostly the same as `include`, but forces you to create a (sometimes) weird *dir* tree structure.

> this could be used to (poorly) implement functions and break the not Turing complete guarantee

Uhm. but how? You can `include` a single file only *once*. I.e. you can't *invoke* `do_smth.meson.build` multiple times:
```meson
include(`do_smth.meson.build`)
...
include(`do_smth.meson.build`) # errors out
```

**Comment by dcbaker** [comment]
Okay, at least if we forbid calling include on the same file multiple times I can live with it. If there isn't one, there should be a test for the include twice case.

I'm still not thrilled, at least with a single 5000 line file I can easily figure out where a variables come from and when it's set without having to jump between dozens of files, but I guess that's a personal preference.

**Comment by TheQwertiest** [comment]
> Okay, at least if we forbid calling include on the same file multiple times I can live with it. 

From the first post...:
"The difference from the previous PR is that this PR forbids multiple inclusions of the same file."

> If there isn't one, there should be a test for the include twice case.

From this PR:
https://github.com/mesonbuild/meson/pull/5209/files#diff-0c5b574484239ba123a7066cea748592

**Comment by TheQwertiest** [comment]
Other possible limitations that could be (should be?) enforced:
- Forbid including `meson.build`.
- Forbid including files from outside of `meson.source_root`.

File: mesonbuild/interpreter.py [hunk 1]
```diff
@@ -3698,6 +3701,55 @@ def func_configure_file(self, node, args, kwargs):
...
+        if os.path.isabs(args[0]):
+            absname = os.path.abspath(args[0])
+        else:
+            absname = os.path.abspath(os.path.join(self.environment.get_source_dir(), self.subdir, args[0]))
+        if not os.path.basename(absname).endswith('.meson.include'):
+            raise InvalidArguments('Name of the included file must follow "*.meson.include" pattern.')
+        if os.path.basename(absname) == environment.build_filename:
+            raise InvalidArguments('Tried to include "%s" - use subdir() instead.' % environment.build_filename)
+        if not self.environment.get_source_dir() in absname
+            raise InvalidArguments('Tried to include file "%s", which is located outside of source directory.' % args[0])
```
Comment (ignatenkobrain) [inline mesonbuild/interpreter.py]: __[Flake8]__ [E113] unexpected indentation ([link](https://sider.review/gh/repos/19784232/pulls/5209#issue-4143770))

<sub>posted by Sider</sub>

File: mesonbuild/interpreter.py [hunk 2]
```diff
@@ -3698,6 +3701,55 @@ def func_configure_file(self, node, args, kwargs):
...
+        if not self.subdir.startswith(self.subproject_dir) and 
+                os.path.join(self.environment.get_source_dir(), self.subproject_dir) in absname:
```
Comment (ignatenkobrain) [inline mesonbuild/interpreter.py]: __[Flake8]__ [E113] unexpected indentation ([link](https://sider.review/gh/repos/19784232/pulls/5209#issue-4143772))

<sub>posted by Sider</sub>

File: mesonbuild/interpreter.py [hunk 3]
```diff
@@ -3698,6 +3701,55 @@ def func_configure_file(self, node, args, kwargs):
...
+            absname = os.path.abspath(args[0])
+        else:
+            absname = os.path.abspath(os.path.join(self.environment.get_source_dir(), self.subdir, args[0]))
+        if not os.path.basename(absname).endswith('.meson.include'):
+            raise InvalidArguments('Name of the included file must follow "*.meson.include" pattern.')
+        if os.path.basename(absname) == environment.build_filename:
+            raise InvalidArguments('Tried to include "%s" - use subdir() instead.' % environment.build_filename)
+        if not self.environment.get_source_dir() in absname
+            raise InvalidArguments('Tried to include file "%s", which is located outside of source directory.' % args[0])
+        if not self.subdir.startswith(self.subproject_dir) and
```
Comment (ignatenkobrain) [inline mesonbuild/interpreter.py]: __[Flake8]__ [W291] trailing whitespace ([link](https://sider.review/gh/repos/19784232/pulls/5209#issue-4143771))

<sub>posted by Sider</sub>

File: mesonbuild/interpreter.py [hunk 4]
```diff
@@ -3698,6 +3701,55 @@ def func_configure_file(self, node, args, kwargs):
...
+        mesonlib.check_direntry_issues(args)
+        if os.path.isabs(args[0]):
+            absname = os.path.abspath(args[0])
+        else:
+            absname = os.path.abspath(os.path.join(self.environment.get_source_dir(), self.subdir, args[0]))
+        if not os.path.basename(absname).endswith('.meson.include'):
+            raise InvalidArguments('Name of the included file must follow "*.meson.include" pattern.')
+        if os.path.basename(absname) == environment.build_filename:
+            raise InvalidArguments('Tried to include "%s" - use subdir() instead.' % environment.build_filename)
+        if not self.environment.get_source_dir() in absname
```
Comment (ignatenkobrain) [inline mesonbuild/interpreter.py]: __[Flake8]__ [E999] SyntaxError: invalid syntax ([link](https://sider.review/gh/repos/19784232/pulls/5209#issue-4143769))

<sub>posted by Sider</sub>

**Comment by TheQwertiest** [comment]
Added the following checks:
- File extension validation: file must not be a `meson.build` and it must have a `meson.include` extension.
- Path validation: file must be in source directory and must not be in subprojects directory (unless current directory is in subproject).

File: mesonbuild/interpreter.py [hunk 5]
```diff
@@ -3698,6 +3701,56 @@ def func_configure_file(self, node, args, kwargs):
...
+        else:
+            absname = os.path.abspath(os.path.join(self.environment.get_source_dir(), self.subdir, args[0]))
+        if not os.path.basename(absname).endswith('.meson.include'):
+            print(os.path.basename(absname))
+            raise InvalidArguments('Name of the included file must follow "*.meson.include" pattern.')
+        if os.path.basename(absname) == environment.build_filename:
+            raise InvalidArguments('Tried to include "%s" - use subdir() instead.' % environment.build_filename)
+        if not self.environment.get_source_dir() in absname:
+            raise InvalidArguments('Tried to include file "%s", which is located outside of source directory.' % args[0])
+        if (not self.subdir.startswith(self.subproject_dir) and
```
Comment (ignatenkobrain) [inline mesonbuild/interpreter.py]: __[Flake8]__ [W291] trailing whitespace ([link](https://sider.review/gh/repos/19784232/pulls/5209#issue-4143773))

<sub>posted by Sider</sub>

**Comment by TheQwertiest** [comment]
Any help with the __[file naming](https://github.com/mesonbuild/meson/pull/5209#issuecomment-479446496)__ would be greatly appreciated!

**Comment by xclaesse** [comment]
Since this is still open for debate, I would like to suggest extra restrictions:
- The file must be in current directory, i.e. forbid `/` in the include() argument. The reason is because it would create a mess with relative paths. If I do `include('../something.meson')` and there I do `subdir('foo')`, `foo` is relative to the `meson.build` or `something.meson`? If it's relative to included file then moving code from `meson.build` to an included file would have to change all relative paths which would be error prone. Also it means files can be all over the place, becoming an chain of includes hard to follow. If the goal is only the split a big file, then having the small pieces in the same dir should be fine, IMHO.
- The filename should start with `meson*` so they get sorted together with `meson.build` and `meson_options.txt` to make it obvious when looking at a dir where is the build definition in the middle of *.c files. I would suggest even `meson_*.build`.

**Comment by TheQwertiest** [comment]
> Wondering if we should rename meson_options.txt to meson_options.build to keep a common file ext.

IMO a common extension would be better, because it will allow to make a proper file association as well. 

PS: `*.build` does not look like meson, but `*.meson` sure does =)

**Comment by xclaesse** [comment]
> > Wondering if we should rename meson_options.txt to meson_options.build to keep a common file ext.
> 
> IMO a common extension would be better, because it will allow to make a proper file association as well.

Most people don't sort files by type [citation needed]. I like how meson.build and meson_options.txt are next each other when I open a meson project, really easy to click on both.

> PS: `*.build` does not look like meson, but `*.meson` sure does =)

Sadly it is about 10 years too late to change `meson.build` to `build.meson` (which would make so much sense, like build.ninja). So we are stuck with .build extension, or maybe *.meson.build? Getting that extention stuff sorted out is very important to get syntax highlight in the files we include. Can most IDE detect syntax only on the extention, or any wildcard (e.g. meson*) ?

**Comment by xclaesse** [comment]
I guess, at least on Linux, the answer is in /usr/share/mime/text/x-meson.xml:
```
  <glob pattern="meson.build"/>
  <glob pattern="meson_options.txt"/>
```

**Comment by TheQwertiest** [comment]
> Can most IDE detect syntax only on the extention, or any wildcard (e.g. meson*) ?

They can (at least those that I use), but OS (Windows that is) can't.

Anyway, extension clean up is nice to have, but it's not the main point of this PR =)

**Comment by tp-m** [comment]
>  I would like to suggest extra restrictions:
> 
>     * The file must be in current directory, i.e. forbid `/` in the include() argument. The reason is because it would create a mess with relative paths. If I do `include('../something.meson')` and there I do `subdir('foo')`, `foo` is relative to the `meson.build` or `something.meson`? If it's relative to included file then moving code from `meson.build` to an included file would have to change all relative paths which would be error prone. Also it means files can be all over the place, becoming an chain of includes hard to follow. If the goal is only the split a big file, then having the small pieces in the same dir should be fine, IMHO.
> 
>     * The filename should start with `meson*` so they get sorted together with `meson.build` and `meson_options.txt` to make it obvious when looking at a dir where is the build definition in the middle of *.c files. I would suggest even `meson_*.build`.

These seem like unnecessary limitations with little benefit to me.

They would prevent me from using this feature,  since it would force me to litter my top-level directory with extra files.

I'm fine with not allowing relative paths, or nested includes, but I don't think there is anything problematic or ambiguous about

```meson
include('meson-snippets/snippet1.meson')
include('meson-snippets/snippet2.meson')
```

for example.

**Comment by TheQwertiest** [comment]
Spring cleanup

**Comment by tp-m** [comment]
:cry:  This is in the top 2 of things I'd like to see added most in Meson for me.

**Comment by TheQwertiest** [comment]
@tp-m , judging by all comments by meson devs in this PR and on IRC, I seriously doubt that this PR (or similar) will ever get merged... But anyone can fork this PR and try again, since I won't be deleting the PR branch =)
It's just that I don't have any motivation (or reason) to maintain this PR anymore.

**Comment by xclaesse** [comment]
AFAIK the only real debate here is about forcing, or not, included files to be in the same directory. I personally still think we should because relative paths are ambiguous otherwise. As far as I understand this PR, paths are relative to the `meson.build` file and not relative to the included file, which makes sense, but can quickly become super weird if you do `subdir()` in that included file, or `find_program('myscript.py')` and that script cannot be in `meson-snippets` folder.

That being said, I'm not going to block this PR on that if others think it's fine.

@tp-m you made me currious, what's your top 1 thing then? :D

**Comment by bonzini** [comment]
> AFAIK the only real debate here is about forcing, or not, included files to be in the same directory. I personally still think we should because relative paths are ambiguous otherwise

I have a usecase for including files in another directory. I have a complex DSL which can't be processed with a single `generator()`, because there are multiple passes needed (a custom tool first, then `dtrace`). Right now I do:

    # other subdirs prepare arrays consumed by trace/meson.build
    foreach f: trace_files
      ...
    endforeach
    trace = static_library('...')
    subdir('util')
    subdir('some-tool')

where some-tool is basically an `executable(..., link_with: util)`. If I want to add tracing support to some-tool, I have to do

    # some-tool/meson.build adds to trace_files and collects sources in a variable
    subdir('some-tool')
    foreach f: trace_files
      ...
    endforeach
    trace = static_library('...')
    ...
    subdir('util')
    # some-tool/meson.build has been processed already, executable
    # must move to main meson.build
    executable('some-tool', sources: some_tool_sources, link_with: trace, util)

but it would be nicer to do

    # some-tool/meson.codegen.build only adds to trace_files
    include('some-tool/meson.codegen.build')
    foreach f: trace_files
      ...
    endforeach
    trace = static_library('...')
    subdir('util')
    subdir('some-tool')
